### ISSUE TITLE
A generic class problem

### ISSUE THREAD

**Issue by cleoold** [microsoft/pyright#2701]
**Describe the bug**
A function with generic parameter does not seem to fit well into an instance method with bound type parameter.

**To Reproduce**
VVV

**Expected behavior**
`identity` should be assignable

**Screenshots or Code**
```python
from typing import Callable, Generic, Iterable, TypeVar

T = TypeVar('T')
U = TypeVar('U')

def identity(x: U) -> U:
    return x

class Test(Generic[T]):
    def fun(self, x: Iterable[T], f: Callable[[T], T]):
        ...

    def caller(self, x: Iterable[T]):
        f: Callable[[T], T] = identity
        self.fun(x, identity)
                       ^
error: Argument of type "(x: U@identity) -> U@identity" cannot be assigned to parameter "f" of type "(T@Test) -> T@Test" in function "fun"
    Type "(x: U@identity) -> U@identity" cannot be assigned to type "(T@Test) -> T@Test"
```

**VS Code extension or command-line**
pyright 1.1.195

**Additional context**
Changing `Iterable[T]` to just `T` the error disappears.

**Comment by erictraut** [comment]
The issue here is that normally TypeVars are solved for a single scope. For example, when solving the TypeVars for `self.fun(x, identity)`, the type checker is solving for the TypeVars in the scope of the `fun` method. This code sample requires solving the TypeVars for the `identity` scope at the same time. In general, pyright doesn't support this, but it might be feasible to support this limited case. There are a few similar cases that pyright already handles as special cases, and it appears that mypy has some special-cased logic to handle this as well.

**Comment by erictraut** [comment]
This will be fixed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.199, which I just published. It will also be included in the next release of pylance.
